### PR TITLE
FIX: Aligning frequency bins to y-axis

### LINKED ISSUE #6929: BUG: Misalignement plotting TFR with linear y-axis
### Describe the bug
Plotting TFRs with linear y-axis, bins for the particular frequencies are not properly aligned with the y-axis.

#### Steps and/or code to reproduce
[This gist](https://gist.github.com/stfnrpplngr/2378c70e6285322f38d54fa04d1fc764) is an example to produce the erroneous plot. The center of the notch filter appears not at 50 Hz regarding the plot.

#### Expected results
50 Hz frequency bin is centered at the 50 Hz y-tick:
Issue_freq_shift_notch_solved
issue_freq_shift_solved2

#### Actual results
50 Hz frequency bin is not centered at the 50 Hz y-tick:
Issue_freq_shift_notch
Issue_freq_shift_notch_2

#### Additional information
Platform:      Windows-10-10.0.16299-SP0
Python:        3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
Executable:    C:\Continuum\Anaconda3\envs\general\python.exe
CPU:           Intel64 Family 6 Model 142 Stepping 10, GenuineIntel: 8 cores
Memory:        Unavailable (requires "psutil" package)
mne:           0.20.dev0
numpy:         1.16.5 {blas=mkl_rt, lapack=mkl_rt}
scipy:         1.3.1
matplotlib:    3.1.1 {backend=Qt5Agg}

sklearn:       0.21.3
numba:         Not found
nibabel:       Not found
cupy:          Not found
pandas:        0.25.1
dipy:          Not found
mayavi:        Not found
pyvista:       Not found
vtk:           Not found

### PR DESCRIPTION
**Pull request by stfnrpplngr**
#### Reference issue
Fixes #6929

#### What does this implement/fix?
Adjusting `extent` using `imshow` for linear y-scale. Now, frequency bins appear centered to the y-ticks.

### CODE REVIEW

File: mne/viz/utils.py [hunk 1]
```diff
@@ -2858,7 +2858,7 @@ def _plot_masked_image(ax, data, times, mask=None, yvals=None,
     else:
         # imshow for linear because the y ticks are nicer
         # and the masked areas look better
-        extent = [times[0], times[-1], yvals[0], yvals[-1] + 1]
+        extent = [times[0], times[-1], yvals[0], yvals[-1]]
```
Comment (larsoner) [inline mne/viz/utils.py]: I think this is still wrong, I get:
Figure_2
```suggestion
        ds = np.array([np.median(np.diff(times)),
                       np.median(np.diff(yvals))]) / 2.
        extent = [times[0] - ds[0], times[-1] + ds[0],
                  yvals[0] - ds[1], yvals[-1] + ds[1]]
```
And later:
```
        time_lims = [extent[0], extent[1]]
        if ylim is None:
            ylim = [extent[2], extent[3]]
```
Gives:

Figure_2

File: mne/viz/utils.py [hunk 2]
```diff
@@ -2875,7 +2875,7 @@ def _plot_masked_image(ax, data, times, mask=None, yvals=None,
                        linewidths=[.75], corner_mask=False,
                        antialiased=False, levels=[.5])
         time_lims = times[[0, -1]]
-        ylim = yvals[0], yvals[-1] + 1
+        ylim = yvals[0], yvals[-1]
```
Comment (larsoner) [inline mne/viz/utils.py]:
```suggestion
        time_lims = [extent[0], extent[1]]
        if ylim is None:
            ylim = [extent[2], extent[3]]
```

File: mne/viz/utils.py (hunk 1 above)
Comment (drammock) [reply mne/viz/utils.py]: ha!  you beat me to it.  I was about to suggest the same solution.

**Comment by agramfort** [comment]
now I see the fix ! merge when you're happy. thx

>

**Comment by stfnrpplngr** [comment]
I think there is nothing more to add. Thanks for the correction @larsoner

File: mne/viz/utils.py [hunk 3]
```diff
@@ -2875,7 +2878,9 @@ def _plot_masked_image(ax, data, times, mask=None, yvals=None,
                        linewidths=[.75], corner_mask=False,
                        antialiased=False, levels=[.5])
         time_lims = times[[0, -1]]
-        ylim = yvals[0], yvals[-1] + 1
+        time_lims = [extent[0], extent[1]]
```
Comment (drammock) [inline mne/viz/utils.py]: forgot to delete the first definition of `time_lims`

Comment (stfnrpplngr) [reply mne/viz/utils.py]: Corrected.

**Comment by drammock** [comment]
CIs are not happy @stfnrpplngr.  OK if I wait to review again until you resolve those issues?

**Comment by stfnrpplngr** [comment]
Ok, I will check what is going wrong.

File: mne/viz/utils.py [hunk 4]
```diff
@@ -2858,7 +2858,10 @@ def _plot_masked_image(ax, data, times, mask=None, yvals=None,
     else:
         # imshow for linear because the y ticks are nicer
         # and the masked areas look better
-        extent = [times[0], times[-1], yvals[0], yvals[-1] + 1]
+        ds = np.array([np.median(np.diff(times)),
+                       np.median(np.diff(yvals))]) / 2.
```
Comment (larsoner) [inline mne/viz/utils.py]: The error is:
```
mne/viz/utils.py:2862: in _plot_masked_image
    np.median(np.diff(yvals))]) / 2.
<__array_function__ internals>:6: in median
    ???
../../../miniconda/lib/python3.7/site-packages/numpy/lib/function_base.py:3502: in median
    overwrite_input=overwrite_input)
../../../miniconda/lib/python3.7/site-packages/numpy/lib/function_base.py:3410: in _ureduce
    r = func(a, **kwargs)
../../../miniconda/lib/python3.7/site-packages/numpy/lib/function_base.py:3561: in _median
    return mean(part[indexer], axis=axis, out=out)
<__array_function__ internals>:6: in mean
    ???
../../../miniconda/lib/python3.7/site-packages/numpy/core/fromnumeric.py:3257: in mean
    out=out, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = array([], dtype=int64), axis = 0, dtype = None, out = None, keepdims = False
    def _mean(a, axis=None, dtype=None, out=None, keepdims=False):
        arr = asanyarray(a)
    
        is_float16_result = False
        rcount = _count_reduce_items(arr, axis)
        # Make this warning show up first
        if rcount == 0:
>           warnings.warn("Mean of empty slice.", RuntimeWarning, stacklevel=2)
E           RuntimeWarning: Mean of empty slice.
```
You need to have a conditional protecting each of these like:
```
if len(times) > 1:
    dt = np.median(np.diff(times)) / 2.
else:
    dt = 0.1
if len(yvals) > 1:
    dy = np.median(np.diff(yvals))
else:
    dy = 0.5
```

**Comment by larsoner** [comment]
@drammock fixed, feel free to review and merge if you're happy, then backport to 0.19

**Comment by drammock** [comment]
just waiting for Travis at this point.

**Comment by drammock** [comment]
Travis failure is unrelated (physionet download timeout).

**Comment by drammock** [comment]
backported

**Comment by stfnrpplngr** [comment]
@drammock 
I think the merge (#6942) was a bit fast. Yesterday, I was still examining the effect of the changes. The output of [test_evoked.py](https://github.com/mne-tools/mne-python/blob/master/mne/tests/test_evoked.py) is wrong now. 
test_plot_evoked_image_issue

The ticks are not aligned anymore which is an easy fix. But I am not sure why the data in the upper plot is shifted...

Shall I open a new Issue for that?

**Pull request merged**